### PR TITLE
Compiler: improve jsoo link in presence of source-maps

### PR DESCRIPTION
**Pull request by hhugo**

### CODE REVIEW

**Comment by hhugo** [comment]
@TyOverby, would you mind reviewing this PR ?

**Comment by TyOverby** [comment]
I’ll do this on Monday

**Comment by pmwhite** [comment]
@TyOverby has deferred review of this PR to me.

File: compiler/lib/link_js.ml [hunk 1]
```diff
@@ -22,27 +22,34 @@ let sourceMappingURL = "//# sourceMappingURL="
...
   in
-  match s with
-  | `Other -> `Other
-  | `Json_base64 base64 ->
-      `Source_map (Source_map_io.of_string (Base64.decode_exn base64))
-  | `Url _ when not resolve_sourcemap_url -> `Drop
-  | `Url url ->
+  match prefix_kind, drop_source_map with
+  | `Other, _ -> Keep
+  | _, true -> Drop
```
[the remark this answers is not in the document]
Comment (hhugo) [reply compiler/lib/link_js.ml]: I've added the change

File: compiler/lib/source_map.ml [hunk 2]
```diff
@@ -211,15 +211,15 @@ let merge = function
         | (gen_line_offset, _, sm) :: rest ->
             let acc =
               { acc with
-                sources = acc.sources @ sm.sources
-              ; names = acc.names @ sm.names
+                sources = List.rev_append sm.sources acc.sources
```
[the remark this answers is not in the document]
Comment (hhugo) [reply compiler/lib/source_map.ml]: > In general, `as @ bs` is not necessarily the same as `rev (rev_append bs as)`

What this PR does in practice is `rev (rev_append bs (rev as))`

All lists accumulated during the merge are in reverse order and reversed back to normal at the end.
I've tried to change the code to make this more obvious.

Also, there is a test for Source_map.merge in compiler/tests-compiler/sourcemap.ml

[the remark this answers is not in the document]
Comment (pmwhite) [reply compiler/lib/source_map.ml]: Excellent, I like the change. Thank you.

**Review by pmwhite — APPROVED** [review]

**Review by pmwhite — APPROVED** [review]

File: compiler/lib/link_js.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (pmwhite) [reply compiler/lib/link_js.ml]: I'm not sure how to resolve these conversations so that merging is unblocked.

**Comment by pmwhite** [comment]
The patch looks good, but I'm trying to run all our internal tests as a sanity check. The compiler tests are failing for what seems to be an unrelated reason; do you have any clue what would cause test diffs like the following:
```
-         if(! param[2])switch$0 = 1;
-        if(! switch$0)
-         {var _c_=param[2],switch$1=0;
-          if(! _c_ || _c_[1])switch$1 = 1;
-          if(switch$1)return 4}}
-      return 1} |}]
+          /*<<test.js 14 9>>*/ if(! param[2]) /*<<test.js 14 23>>*/ switch$0 = 1;
+         /*<<test.js 15 8>>*/ if(! switch$0)
+          /*<<test.js 16 9>>*/ { /*<<test.js 16 10>>*/ var
+            /*<<test.js 16 17>>*/ _c_=param[2],
+            /*<<test.js 16 35>>*/ switch$1=0;
+           /*<<test.js 17 10>>*/ if(! _c_ || _c_[1])
+            /*<<test.js 17 29>>*/ switch$1 = 1;
+           /*<<test.js 18 10>>*/ if(switch$1) /*<<test.js 18 22>>*/ return 4}}
+       /*<<test.js 19 6>>*/ return 1 /*<<test.js 19 14>>*/ } |}]
```

**Comment by hhugo** [comment]
> param[2]) /*<<test.js 14 23>>*/ switch$0 = 1;
> +         /*<<test.js 15 8>>*/ if(! switch$0)
> +          /*<<test.js 16 9>>*/ { /*<<test.js 16 10>>*/ var
> +            /*<<test.js 16 17>>*/ _c_=param[2],
> +            /*<<test.js 16 35>>*/ switch$1=0;
> +           /*<<test.js 17 10>>*/ if(! _c_ || _c_[1])
> +            /*<<test.js 17 29>>*/ switch$1 = 1;
> +           /*<<test.js 18 10>>*/ if(switch$1) /*<<test.js 18 22>>*/ return 4}}
> +       /*<<test.js 19 6>>*/ return 1 /*<<test.js 19 14>>*/ } |}]

The diff looks like you've turned `--debuginfo` on, somewhere.

**Comment by pmwhite** [comment]
> The diff looks like you've turned `--debuginfo` on, somewhere.

Aha, thank you.

**Comment by pmwhite** [comment]
FYI: I've done some basic internal testing, to make sure this didn't break anything unexpectedly. Everything seems good to me, so go ahead and merge this.

**Pull request merged**